### ISSUE TITLE
Conflict when adding `coder_metadata` with duplicate `resouce_id`s

### ISSUE THREAD

**Issue by matifali** [coder/terraform-provider-coder#305]
**Problem Description**

1. While adding `coder_metadata` blocks inside a module we need to pass the resource_id to link the metadata to. The issue arises when the `resource_id` of the instance is linked to an existing `coder_metadata` block in the main template, which causes a conflict and the template fails to build.

2. Attempting to use an alternative identifier, such as `agent_id`, doesn't show metadata on the workspace page.

**Steps to Reproduce**  
1. Add a `coder_metadata` block to a module and link it to the `resource_id` of an instance.  
2. Add another `coder_metadata` block in the main template referencing the same `resource_id`.  
3. Observe the conflict in the failed build.  
4. Attempt to use `agent_id` instead in the module and note the absence of display metadata.  

**Impact**  
- Limits the flexibility of the coder modules because of possible conflicts.

**Next Steps**  
- Update the handling of `coder_metadata` to avoid conflicts between module-level and main template declarations when sharing the same `resource_id`.  One possible solution is to merge and append the new [`item`](https://registry.terraform.io/providers/coder/coder/latest/docs/resources/metadata#nestedblock--item) entries when we detect a conflict in `resource_id`.
- Investigate why the `agent_id` does not work with `coder_metadata`.

> [!Note]
> We did allow adding multiple `coder_metadata` with the same `resource_id`. See coder/coder#6517.

**Comment by ethanndickson** [comment]
#1 sounds like a good solution, just a fair bit more work. 
#2 would effectively be merging the agent's static metadata with the first resource of the workspace, which would be equally confusing as merging it with the dynamic agent metadata.

**Comment by matifali** [comment]
> 2. This would effectively be merging the agent's static metadata with the first resource of the workspace, which would be equally confusing as merging it with the dynamic agent metadata.

I don't mean we merge the `coder_metadata` assigned to different resources but merge if multiple `coder_metadata` are linked to a single resource. Consider the example below,

```tf
resource "coder_metadata" "module" {
  count       = data.coder_workspace.me.start_count
  resource_id = resource.aws_instnace.id

  item {
    key   = "username"
    value = module.amazon-dcv-windows[count.index].username
  }
  item {
    key       = "password"
    value     = module.amazon-dcv-windows[count.index].password
    sensitive = true
  }
}

resource "coder_metadata" "template" {
  count       = data.coder_workspace.me.start_count
  resource_id = resource.aws_instnace.id # note that both use the same reosurce_id

item {
    key   = "region"
    value = data.coder_parameter.region.value
  }
  item {
    key   = "instance type"
    value = aws_instance.dev.instance_type
  }
}
```

We should display a single coder_metadata section and merge the `item` from both `coder_metadata` resources as they are both linked with the same `resource.aws_instance.dev` resource. We may still get conflicts if the `item` keys are not unique.

What do you think about it @bpmct?

**Comment by ethanndickson** [comment]
Oh I see what you mean, yeah, that was exactly what I proposed in my original comment.

**Comment by matifali** [comment]
> Oh I see what you mean, yeah. That's what you suggested originally, I definitely agree that's a good solution. but that doesn't address where coder_metadata attached to a coder_agent should go.

Yes, maybe we can narrow the scope so as not to support the `coder_agent` for now, just update our modules to pass another resource_id as input, and attach all metadata to that resource from within the modules.
